README.md
Go Buckytools
==============

BEWARE!
This is my ugly hack of @jjneely's Buckytools!

Original repo is https://github.com/jjneely/buckytools

I just moved support of jump_hash to master branch and backported Jack's fixes.

For using jump_fnv1a hash use `--hash jump_fnv1a` cli parameter of buckyd.
Please note that instances are ignored in this mode (contrary of carbon-c-relay),
so please sort your hosts in ascending order to get compatible results!

If you're using sparse whisper files - do not forget to add `--sparse` to buckyd too.

I'm planning to create PR in original repo and remove this hack.

Also debian packaging is broken in this version, please do not use it.

-- @deniszh

---

Tools for managing large consistent hashing Graphite clusters.

Buckminster (Bucky) Fuller once said,

> If you want to teach people a new way of thinking, don’t bother trying to teach
> them. Instead, give them a tool, the use of which will lead to new ways of
> thinking.

Carbon molecules known as fullerenes were also named after Bucky due
to their geodesic arrangement of atoms.  So, this is my contribution
to the Graphite naming scheme.

When working large consistent hashing Graphite clusters doing simple
maintenance can involve a lot of data moving around.  Normally, one would reach
for the Carbonate tools:

    https://github.com/jssjr/carbonate

These are good tools and I highly recommend them.

However, when the terabytes of WSP files and the number of storage nodes
stack up you start to find scaling problems:

* The python implementation of whisper-fill is slow
* There is no atomic method to rebalance a single metric
  resulting in query inconsistency for possible days after the
  metric was moved
* Handling renames and resulting backfills at scale is still
  difficult

So I wanted to use the speed and concurrency of Go to build more efficient
tools to help manage large consistent hashing cluster.

Tools
=====

These are the tools included and their functionality.

* **bucky-pickle-relay** -- A daemon that accepts and decodes Graphite's
  Pickle protocol and will tcp stream the equivalent plaintext version
  to a carbon-relay.  Useful with [carbon-c-relay][1].
* **bucky-fill** -- A `whisper-fill` compatible utility that is nearly
  an order of magnitude faster.
* **bucky-isempty** -- A utility for discovering WSP databases that
  contain no valid data points.
* **buckyd** -- A daemon for each Graphite node that tracks the
  configuration of the hash ring and exposes a REST API for
  interacting with the raw metric DBs on disk.
* **bucky** -- Command line Graphite cluster manager.  Modules:
  * **backfill** -- Backfill old metrics into new names.
  * **delete** -- Delete metrics via list or regular expression.
  * **du** -- Measure the storage consumed by a list of regular expression of
    metrics.
  * **inconsistent** -- Find metrics that are stored in the wrong server
    according to the hash ring.
  * **json** -- Convert newline separated lists to JSON arrays.
  * **list** -- Discover and verify metrics.
  * **locate** -- Calculate metric locations from the hash ring.
  * **rebalance** -- Move inconsistent metrics to the correct location
    and delete the source immediately after successful backfill.
  * **restore** -- Restore from a tar archive.
  * **severs** -- List each server's known hash ring and verify that
    all hash rings are consistent.
  * **tar** -- Make an archive of a list or regular expression of metric
    names and dump it in tar format to STDOUT.
* **gentestmetrics** -- Command that generates random Graphite style metrics
  to stdout purely for testing.
* **bucky-sparsify** -- Rewrites `.wsp` files into sparse files.

The heavy lifting commands use a set of worker threads to do IO work
which can be set at the command line with -w.

Assumptions
===========

These tools assume the following are true:

* You Graphite carbon-cache servers have one Whisper DB store.  Not multiple
  mount points with carbon-cache configured with their own DB store.
* Your hash ring is set to a `REPLICATION_FACTOR` of 1

These aren't set in stone, just what I was working with as I built the tool.  I
very much hope that some of these will be solved with further development.

Daemon Usage
============

Each data storage node in the Graphite cluster needs a **buckyd** daemon
running as the same user as the other Graphite tools.  I use an Upstart
job to keep mine running.  The important bit here is that you must
pass to the daemon as arguments the members of the consistent hash ring.

    $ cat /etc/init/buckyd.conf
    description "Buckyd Daemon for Managing Graphite Clusters"
    author      "Jack Neely <jjneely@42lines.net>"

    setuid graphite

    exec /path/to/buckyd --node graphite010-g5 \
        -b 192.168.1.1:5678 \
	graphite010-g5:a graphite010-g5:b \
	graphite011-g5:a graphite011-g5:b \
	graphite012-g5:a graphite012-g5:b

Here `--node` is the name of this Graphite node (if different from what
is derived from the host name).  `-b` or `--bind` is the address to bind
to.  You can also specify `--prefix`` where your Whisper data store is
and `--tmpdir` where the daemon can write temporary files.

This exposes a REST API that is documented in REST_API_NOTES.md.

Client Usage
============

The **bucky** tool is self documenting.  You can run:

    bucky help

to see a list of modules and available flags and what arguments are needed.
Detailed help is available by specifying a module name:

    bucky help backfill

Most commands need a `--host` or `-h` flag to specify the initial Graphite
host to connect to where the client will discover the entire hash ring.
You can also set the `BUCKYHOST` environment variable rather than
specify this flag for each command.

Other common flags are:

* `-s` Operate only on the initial Graphite host.
* `-f` Requests the remote daemon to refresh its cache of local metrics.
* `-j` Read from STDIN or dump to STDOUT JSON data rather than text.
* `-r` Regular expression mode.
* `-w` Number of worker threads.

Examples
========

Rebalance a cluster with newly added storage nodes:

    GOMAXPROCS=4 bucky rebalance -h graphite010-g5:4242 \
        -w 100 2>&1 | tee rebalance.log

Discover the exact storage used by a set of metrics:

    export BUCKYHOST=-h graphite010-g5:4242
    bucky du -r '^1min\.ipvs\.'

To Do / Bugs
============

* Authentication -- Negotiate and Kerberos support.  Probably Basic as well.
* Code clean up -- I wrote a lot of this quickly, it needs love in a
  lot of places.
* tests
* Make all modules aware of possible duplicate metrics.
* Speed testing and improvements.
* Move the lower level GET, POST, DELETE, HEAD functions into a single
  common file/place.
* Retries
* Rebalance needs to optionally be aware of machines not in the hash ring that
  the rebalance should vacate.
* Graceful restarts and shutdowns?  https://github.com/facebookgo/grace
* graphite-project/carbon's master branch contains this change:

    https://github.com/graphite-project/carbon/commit/024f9e67ca47619438951c59154c0dec0b

  This will cause a few metrics to be assigned a different position in the
  hash ring.  We need to account for this algorithm change somehow.

[1]: https://github.com/grobian/carbon-c-relay


debian/changelog
buckytools (0.1.7-4) unstable; urgency=medium

  * jump hash
  * backport of fixes from master
  * bucky-sparsify
  * sparse support

 -- Jack Neely <jjneely@42lines.net>  Thu, 29 Oct 2015 15:26:41 -0400

buckytools (0.1.6-1) unstable; urgency=medium

  * bucky-pickle-relay logging fixes
  * bucky-pickle-relay decodes pickles in the connection handler rather than
    a dedicated goroutine.

 -- Jack Neely <jjneely@42lines.net>  Thu, 29 Oct 2015 15:26:41 -0400

buckytools (0.1.5-1) unstable; urgency=medium

  * Refactor bucky-pickle-relay to be more idiomatic Go and perform
    better.
  * Add signal handling and instrumentation to bucky-pickle-relay
  * Do not re-use http.Request objects in the bucky client
  * Certify Go 1.5 safe

 -- Jack Neely <jjneely@42lines.net>  Wed, 28 Oct 2015 16:14:08 -0400

buckytools (0.1.4-1) unstable; urgency=medium

  * Rebuild with Go 1.5.1
  * Build with autobuilders
  * Include bucky stat command

 -- Jack Neely <jjneely@42lines.net>  Thu, 22 Oct 2015 15:52:49 -0400

buckytools (0.1.4) unstable; urgency=medium

  * bucky-pickle-relay: Improve logging
  * bucky-pickle-relay: Separate timeouts from the receive operation and the
    send operation.  We now wait 5 minutes before idling out a client's
    connection.

 -- Jack Neely <jjneely@42lines.net>  Tue, 08 Sep 2015 11:34:19 -0400

buckytools (0.1.3) unstable; urgency=medium

  * Except *big.Int timestamps/data values types

 -- Jack Neely <jjneely@42lines.net>  Thu, 03 Sep 2015 15:11:23 -0400

buckytools (0.1.2) unstable; urgency=medium

  * bucky-pickle-relay: accept float64 timestamps

 -- Jack Neely <jjneely@42lines.net>  Tue, 04 Aug 2015 14:12:30 -0400

buckytools (0.1.1) unstable; urgency=medium

  * bucky-pickle-relay: Handle TCP RST packets better

 -- Jack Neely <jjneely@42lines.net>  Thu, 30 Jul 2015 14:43:11 -0400

buckytools (0.1.0) unstable; urgency=medium

  * Build of 0.1.0 with bucky-pickle-relay

 -- Jack Neely <jjneely@42lines.net>  Fri, 24 Jul 2015 15:57:49 -0400

buckytools (0.0.5) unstable; urgency=medium

  * Version bump for dh-golang packaging

 -- Jack Neely <jjneely@42lines.net>  Mon, 06 Jul 2015 16:18:52 -0400

buckytools (0.0.4) unstable; urgency=medium

  * Bugfixes and locking support

 -- Jack Neely <jjneely@42lines.net>  Fri, 12 Jun 2015 14:50:57 -0400

buckytools (0.0.3) unstable; urgency=low

  * Version bump.

 -- Jack Neely <jjneely@42lines.net>  Fri, 15 May 2015 18:18:03 +0000

buckytools (0.0.2) unstable; urgency=low

  * Version bump.

 -- Jack Neely <jjneely@42lines.net>  Fri, 08 May 2015 13:00:00 -0400

buckytools (0.0.1) unstable; urgency=low

  * Initial release/packaging.

 -- Jack Neely <jjneely@42lines.net>  Mon, 04 May 2015 13:19:20 -0400


bucky/common.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"net/url"
	"os"
	"strings"
)

import . "github.com/deniszh/buckytools"

// httpClient is a cached http.Client. Use GetHTTP() to setup and return.
var httpClient *http.Client

// MetricData represents an individual metric and its raw data.
// XXX: Unify this with MetricStatType?
type MetricData struct {
	Name    string
	Size    int64
	Mode    int64
	ModTime int64
	Data    []byte
}

// GetHTTP returns a *http.Client that can be used to interact with remote
// buckyd daemons.
func GetHTTP() *http.Client {
	if httpClient != nil {
		return httpClient
	}

	httpClient = new(http.Client)

	// Set a 30 second timeout on all operations
	//httpClient.Timeout = 30 * time.Second

	return httpClient
}

// DeleteMetric sends a DELETE request for the given metric to the given
// server.  The port is assumed the same for all Bucky daemons in the
// hash ring.
func DeleteMetric(server, metric string) error {
	httpClient := GetHTTP()
	u := &url.URL{
		Scheme: "http",
		Host:   fmt.Sprintf("%s:%s", server, Cluster.Port),
		Path:   "/metrics/" + metric,
	}

	r, err := http.NewRequest("DELETE", u.String(), nil)
	if err != nil {
		log.Printf("Error building request: %s", err)
		return err
	}

	resp, err := httpClient.Do(r)
	if err != nil {
		log.Printf("Error communicating: %s", err)
		return err
	}
	defer resp.Body.Close()

	switch resp.StatusCode {
	case 200:
		log.Printf("DELETED: %s", metric)
	case 404:
		log.Printf("Not found / Not deleted: %s", metric)
		return fmt.Errorf("Metric not found.")
	case 500:
		msg, err := ioutil.ReadAll(resp.Body)
		if err != nil {
			msg = []byte(err.Error())
		}
		log.Printf("Error: Internal Server Error: %s", string(msg))
		return fmt.Errorf("Error: Internal Server Error: %s", string(msg))
	default:
		log.Printf("Error: Unknown response from server.  Code %s", resp.Status)
		return fmt.Errorf("Unknown response from server.  Code %s", resp.Status)
	}

	return nil
}

// GetMetricData retrieves the binary Whisper data for a given metric
// name that lives on the given server.  The port buckyd runs on is
// assumed to be the same as other servers in the hash ring.
func GetMetricData(server, name string) (*MetricData, error) {
	httpClient := GetHTTP()
	u := &url.URL{
		Scheme: "http",
		Host:   fmt.Sprintf("%s:%s", server, Cluster.Port),
		Path:   "/metrics/" + name,
	}
	r, err := http.NewRequest("GET", u.String(), nil)
	if err != nil {
		log.Printf("Error building request: %s", err)
		return nil, err
	}

	resp, err := httpClient.Do(r)
	if err != nil {
		log.Printf("Error downloading metric data: %s", err)
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		body, _ := ioutil.ReadAll(resp.Body)
		log.Printf("Error: Fetching [%s]:%s returned status code: %d  Body: %s",
			server, name, resp.StatusCode, string(body))
		return nil, fmt.Errorf("Fetching metric returned status code: %s", resp.Status)
	}

	data := new(MetricData)
	err = json.Unmarshal([]byte(resp.Header.Get("X-Metric-Stat")), &data)
	if err != nil {
		log.Printf("Error unmarshalling X-Metric-Stat header for [%s]:%s: %s", server, name, err)
		return nil, err
	}

	data.Data, err = ioutil.ReadAll(resp.Body)
	if err != nil {
		log.Printf("Error reading response body: %s", err)
		return nil, err
	}

	return data, nil
}

func StatRemoteMetric(server, metric string) (*MetricStatType, error) {
	httpClient := GetHTTP()
	u := &url.URL{
		Scheme: "http",
		Host:   fmt.Sprintf("%s:%s", server, Cluster.Port),
		Path:   "/metrics/" + metric,
	}

	r, err := http.NewRequest("HEAD", u.String(), nil)
	if err != nil {
		log.Printf("Error building request: %s", err)
		return nil, err
	}

	resp, err := httpClient.Do(r)
	if err != nil {
		log.Printf("Error communicating: %s", err)
		return nil, err
	}
	defer resp.Body.Close()

	switch resp.StatusCode {
	case 200:
		data := resp.Header.Get("X-Metric-Stat")
		if data == "" {
			log.Printf("No stat data returned for: %s", metric)
			return nil, fmt.Errorf("No stat data returned for: %s", metric)
		}
		stat := new(MetricStatType)
		err := json.Unmarshal([]byte(data), &stat)
		if err != nil {
			log.Printf("Error: Could not parse X-Metric-Stat header for %s", metric)
			return nil, err
		}
		return stat, nil
	case 404:
		log.Printf("Metric not found: %s", metric)
		return nil, fmt.Errorf("Metric not found.")
	case 500:
		msg, err := ioutil.ReadAll(resp.Body)
		if err != nil {
			msg = []byte(err.Error())
		}
		log.Printf("Error: Internal Server Error: %s", string(msg))
		return nil, fmt.Errorf("Error: Internal Server Error: %s", string(msg))
	default:
		log.Printf("Error: Unknown response from server.  Code %s", resp.Status)
		return nil, fmt.Errorf("Unknown response from server.  Code %s", resp.Status)
	}

	// shouldn't get here
	return nil, fmt.Errorf("Unexpected error in StatRemoteMetric()")
}

// GetSingleHashRing connects to the given server and returns a JSONRingType
// representing its hashring configuration.  An error value is
// set if we could not retrieve the hashring information.  The server
// string must include any port information.
func GetSingleHashRing(server string) (*JSONRingType, error) {
	u := &url.URL{
		Scheme: "http",
		Host:   server,
		Path:   "/hashring",
	}
	httpClient := GetHTTP()

	r, err := http.NewRequest("GET", u.String(), nil)
	if err != nil {
		log.Printf("Error building request: %s", err)
		return nil, err
	}
	resp, err := httpClient.Do(r)
	if err != nil {
		log.Printf("Error retrieving URL: %s", err)
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		log.Printf("Abort: /hashring API called returned: %s", resp.Status)
		return nil, fmt.Errorf("/hashring API called returned: %s", resp.Status)
	}

	ring := new(JSONRingType)
	blob, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		log.Printf("Error reading response body: %s", err)
		return nil, err
	}
	err = json.Unmarshal(blob, &ring)
	if err != nil {
		log.Printf("Could not unmarshal JSON from host %s: %s", server, err)
		return nil, err
	}

	return ring, nil
}

// HostPort is a convenience variable for sub-commands.  This holds the
// HOST:PORT to connect to if SetupHostname() is called in init()
var HostPort string

// SetupHostname sets up a generic find the host to connect to flag
func SetupHostname(c Command) {
	var host string
	if os.Getenv("BUCKYHOST") != "" {
		host = os.Getenv("BUCKYHOST")
	} else {
		host = "localhost:4242"
	}

	c.Flag.StringVar(&HostPort, "h", host,
		"HOST:PORT to find a remote buckyd daemon. Port is optional.")
	c.Flag.StringVar(&HostPort, "host", host,
		"HOST:PORT to find a remote buckyd daemon. Port is optional.")
}

// SingleHost is a convenience variable for sub-commands.  A sub-command
// must call SetupSingle() from their init() to enable.  This is true if
// -s or --single is present and restricts the operation to the original
// host and not the entire cluster.
var SingleHost bool

// SetupSingle sets up the -s|--single command flag
func SetupSingle(c Command) {
	c.Flag.BoolVar(&SingleHost, "s", false,
		"Operate on the given hostname only, do not discover the cluster members.")
	c.Flag.BoolVar(&SingleHost, "single", false,
		"Operate on the given hostname only, do not discover the cluster members.")
}

// JSONOuput is a convenience variable for sub-commands.  If setup by calling
// SetupJSON() from a sub-command's init() this will be true if the -j or
// --json flags are present and the command should dump out JSON encoded data.
var JSONOutput bool

// SetupJSON installs the -j and --json flags in the given Command
func SetupJSON(c Command) {
	c.Flag.BoolVar(&JSONOutput, "j", false,
		"Instead of text ouput JSON encoded data.")
	c.Flag.BoolVar(&JSONOutput, "json", false,
		"Instead of text ouput JSON encoded data.")
}

// CleanMetric sanitizes the given metric key by removing adjacent "."
// characters and replacing any "/" characters with "."
func CleanMetric(m string) string {
	// Slash isn't really illegal but gets interperated as a directory
	// on the Graphite server
	m = strings.Replace(m, "/", ".", -1)

	// Adjacent "." end up producing non-normalized paths
	for strings.Index(m, "..") != -1 {
		m = strings.Replace(m, "..", ".", -1)
	}

	// Can't begin with a "."
	if strings.Index(m, ".") == 0 {
		m = m[1:]
	}

	return m
}
